iavl/iavl_tree.go

```diff
@@ -41,139 +135,303 @@ func NewIAVLTree(cacheSize int, db dbm.DB) *IAVLTree {
+		// Null root
 		return nil
+
+	} else if t.ndb != nil {
+		firstNode.save(t)
 	}
+
+	last := t.roots.Back()
+	lastNode := last.Value.(*IAVLNode)
+
 	if t.ndb != nil {
-		t.root.save(t)
+		// TODO: should be a loop, if the rootsMax can change
+		if rootsMax > 1 && t.roots.Len()+1 > rootsMax {
+			//fmt.Printf("##### Cleanup!!! ######\n")
+
+			// TODO: should be locking any changes to deletes?
+			t.ndb.deletes = append(t.ndb.deletes, lastNode.hash)
+			t.roots.Remove(last)
+
+			t.ndb.SaveDeletes(t.ndb.batch)
+		}
+
+		//fmt.Printf("Orphan Count %d\n", len(t.ndb.orphans))
+		t.ndb.SaveOrphans(firstNode.hash, t.ndb.orphans)
+		t.ndb.orphans = make([][]byte, 0)
+
+		t.roots.PushFront(firstNode)
+		t.ndb.SaveRoots(t)
+		t.ndb.SaveVersion(t)
+		t.version++
+
 		t.ndb.Commit()
+
+		// TODO: Move to a separate go-routine?
+		t.ndb.Prune()
+
+	} else {
+		t.HashWithCount()
+		if rootsMax > 1 && t.roots.Len()+1 > rootsMax {
+			t.roots.Remove(last)
+			t.roots.PushFront(firstNode)
+			t.version++
+		}
 	}
-	return t.root.hash
+
+	// What's the final status?
+	//t.PrintRoots()
+	root := t.GetRoot(t.version)
+	//fmt.Printf("root = '%X'\n", root.hash)
+
+	return root.hash
 }
 
 // Sets the root node by reading from db.
 // If the hash is empty, then sets root to nil.
 func (t *IAVLTree) Load(hash []byte) {
+	//fmt.Printf("***** Loading a tree\n")
 	if len(hash) == 0 {
-		t.root = nil
+		SetValue(t.roots, 0, nil)
 	} else {
-		t.root = t.ndb.GetNode(t, hash)
+		//fmt.Printf("Loading Root\n")
+		t.ndb.GetVersion(t)
+		t.ndb.GetRoots(t)
+		t.ndb.GetDeletes()
+		SetValue(t.roots, 0, t.ndb.GetNode(t, hash))
 	}
 }
 
+// GetVersion finds the value for the current version
 func (t *IAVLTree) Get(key []byte) (index int, value []byte, exists bool) {
-	if t.root == nil {
+	//fmt.Printf("Get ")
+	root := t.GetRoot(t.version)
+	if root == nil {
+		//fmt.Printf("Failed Get for %d\n", t.version)
+		return 0, nil, false
+	}
+	return root.get(t, key)
+}
+
+// GetVersion finds the value for an older committed version
+func (t *IAVLTree) GetVersion(key []byte, version int) (index int, value []byte, exists bool) {
+	root := t.GetRoot(version)
+	if root == nil {
 		return 0, nil, false
 	}
-	return t.root.get(t, key)
+	return root.get(t, key)
 }
 
+// GetByIndex the ith key and value
 func (t *IAVLTree) GetByIndex(index int) (key []byte, value []byte) {
-	if t.root == nil {
+	root := t.GetRoot(t.version)
+	if root == nil {
 		return nil, nil
 	}
-	return t.root.getByIndex(t, index)
+	return root.getByIndex(t, index)
 }
 
+// Remove a key from the tree
 func (t *IAVLTree) Remove(key []byte) (value []byte, removed bool) {
-	if t.root == nil {
+	//fmt.Printf("Remove ")
+
+	root := t.GetRoot(t.version)
+	if root == nil {
 		return nil, false
 	}
-	newRootHash, newRoot, _, value, removed := t.root.remove(t, key)
+
+	newRootHash, newRoot, _, value, removed := root.remove(t, key)
 	if !removed {
 		return nil, false
 	}
+
 	if newRoot == nil && newRootHash != nil {
```

Comment:
My guess is that it was there because of the rolls. Maybe because the node rolled to the top is no longer in the cache?